### PR TITLE
Changes to work with Cloudant local

### PR DESCRIPTION
**Pull request by rhyshort**
- Make sequence numbers transparent, previously they were expected
  to be strings, this causes failures when replicating with
  Cloudant Local which uses arrays for sequence numbers.

- Also added a more configurable couch config to enable
  testing with any couchDB server.

### CODE REVIEW

File: sync-core/src/test/java/com/cloudant/common/CouchTestBase.java [hunk 1]
```diff
@@ -25,7 +29,10 @@ public CouchConfig getCouchConfig(String db) {
                         " is required to run tests with Cloudant.");
             }
             return config;
-        } else {
+        } else if (SPECIFIED_COUCH) {
+            return SpecifiedCouch.defaultConfig(db);
```
Comment (tomblench) [inline sync-core/src/test/java/com/cloudant/common/CouchTestBase.java]: I don't really see the value of this SpecifiedCouch class. Why not just let the user pass in a url string and have done with it?

Comment (rhyshort) [inline sync-core/src/test/java/com/cloudant/common/CouchTestBase.java]: it follows the pattern we've used previously with cloudant, it has its trade offs, there are a lot more config options, a url would be easier, but may be confusing is you don't know how to pass a username and password through the url

File: sync-core/src/main/java/com/cloudant/mazha/CouchURIHelper.java [hunk 2]
```diff
@@ -61,6 +63,17 @@ public URI changesUri(Map<String, Object> query) {
                 this.rootUriString,
                 "_changes"
         );
+
+        //lets find the since parameter
+        if(query.containsKey("since")){
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/mazha/CouchURIHelper.java]: would it be better to do the JSON encoding at the call site (whichever method is preparing the _changes url) rather than special-casing it here?

Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/mazha/CouchURIHelper.java]: Not really, encoding it here means only the helper needs to know about encoding, where as if we place it at the call site for the method we have the potential for the implementation detail about sequence numbers to leak out more

**Comment by mikerhodes** [comment]
Split 3b9557f in two:

1. `Also added a more configurable couch config to enable  testing with any couchDB server.`
2. The fixes for Local.

"Also" should be a warning light in your head.

Also, we made the sequence number opaque to the code, not transparent. I'd just use "pass-through".

File: sync-core/src/main/java/com/cloudant/mazha/CouchURIHelper.java [hunk 3]
```diff
@@ -61,6 +63,17 @@ public URI changesUri(Map<String, Object> query) {
...
+            Object since = query.get("since");
+            if(!(since instanceof String)){
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/mazha/CouchURIHelper.java]: Did you confirm that the wrong thing happens if we remove this check and pass a string? That is, we end up with `since="123-blah"` rather than `since=123-blah`.

File: sync-core/src/main/java/com/cloudant/mazha/CouchURIHelper.java (hunk 2 above)
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/mazha/CouchURIHelper.java]: Tom might have a point, it's a slightly weird place to have this logic. But given the warped nature of some of this code it might be the most sensible right now. Who calls this method with a `since` param?

File: sync-core/src/test/java/com/cloudant/common/CouchTestBase.java (hunk 1 above)
Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/common/CouchTestBase.java]: Can we replace the `TEST_WITH_CLOUDANT` stuff with `SPECIFIED_COUCH`? It seems that they're both doing the same job, that is allowing the user to configure something other than a localhost-ish couchdb (just CloudantConfig is a bit more specific)?

File: sync-core/src/test/java/com/cloudant/sync/replication/DBWithSlashReplicationTest.java [hunk 4]
```diff
@@ -66,7 +66,7 @@ public void replicationPullAndPushDbWithSlash() throws Exception{
         replicator.start();
         Assert.assertEquals(Replicator.State.STARTED, replicator.getState());
 
-        while(replicator.getState() != Replicator.State.COMPLETE) {
+        while(replicator.getState() != Replicator.State.COMPLETE || replicator.getState() == Replicator.State.ERROR) {
```
Comment (mikerhodes) [inline sync-core/src/test/java/com/cloudant/sync/replication/DBWithSlashReplicationTest.java]: This and the change to `PullReplicatorTest` look like they belong in a separate commit which explains what the change is for.

File: sync-core/src/main/java/com/cloudant/mazha/CouchURIHelper.java (hunk 3 above)
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/mazha/CouchURIHelper.java]: Yeah, I ran a test against couch 1.6 and it was failing because it was encoding some quotes in the url.

File: sync-core/src/main/java/com/cloudant/sync/replication/CouchDB.java [hunk 5]
```diff
@@ -43,8 +43,8 @@
     public String getCheckpoint(String checkpointId);
     public void putCheckpoint(String checkpointId, String sequence);
 
-    public ChangesResult changes(String lastSequence, int limit);
-    public ChangesResult changes(Replication.Filter filter,String lastSequence, int limit);
+    public ChangesResult changes(Object lastSequence, int limit);
+    public ChangesResult changes(Replication.Filter filter,Object lastSequence, int limit);
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/replication/CouchDB.java]: Check your formatting, spaces after `,`.

**Comment by mikerhodes** [comment]
Make the commit messages a little more result-orientated rather than what one did orientated.

Also, any docs need updating with the removal of CloudantConfig?

Code is looking good though, pleased we fixed this quite quickly.

**Comment by rhyshort** [comment]
Yeah I updated the docs in 41e220b

**Comment by rhyshort** [comment]
Verbal +1 from @mikerhodes merging
